hilti/runtime/include/types/stream.h

```diff
@@ -1568,11 +1568,11 @@ class View final {
     }
 
     void _ensureValid() const {
-        //        if ( ! _begin.chain() )
-        //            throw InvalidIterator("view has invalid beginning");
+        if ( ! _begin.isValid() )
+            throw InvalidIterator("view has invalid beginning");
 
-        //        if ( ! _begin.isValid() )
-        //            throw InvalidIterator("view has invalid beginning");
+        if ( (! _begin.isUnset()) && _begin.offset() < _begin.chain()->offset() )
+            throw InvalidIterator("view starts before available range");
```

Comment:
I am not a fan of the error message since for anybody but direct users of the C++ API they might have trouble figuring out what change they need to make to address the issue.

I am wondering whether it would make more sense to instead push this check down into `View::extract` where we would have a chance to raise an exception similar to the one we emit in the parser builder if less than expected `&size` bytes could be extracted.

Reply:
Agree in principle, but not sure how to make it much better. Even though `extract` is what the ticket is about, I think the problem doesn't just apply there but also to other places requiring a valid start iterator, hence adding it to `_ensureValid`. And we already have a quite a few other exceptions thrown by the runtime that are probably similarly challenging from a grammar perspective, which is a result of how our layering.